Python 3

ALescoulie opened the pull request:
Replaced yield tests. Removed type annotations. Added more compatibility code for python 2.

ALescoulie commented:
Just realized there was a small error in formatting a string in conifg, fixed it in a commit.

ALescoulie commented:
With the fix to GromacsWrapper only 10 tests of 73 are failing, Solvate wet octanol, and the two analysis tests

ALescoulie commented:
With the fix to GromacsWrapper only 10 tests of 73 are failing, solvation of wet octanol, and the two analysis tests

ALescoulie commented on .github/workflows/CI.yml:
```diff
@@ -0,0 +1,87 @@
+      uses: codecov/codecov-action@v1
+      with:
+        file: ./coverage.xml
+        fail_ci_if_error: true
```
Started CI file similar to one by @orbeckst for gromacswrapper, work in progress.

ALescoulie commented on mdpow/equil.py:
```diff
@@ -206,10 +211,23 @@ def __init__(self, molecule=None, **kwargs):
         super(Simulation, self).__init__(**kwargs)
 
+    @staticmethod
+    def load_pickle():
```
For compatibility between 2 and 3, could probably also use six

orbeckst replied:
six is preferred — better use existing, tested libraries and six is standard and already used

ALescoulie commented on mdpow/tests/test_forcefields.py:
```diff
@@ -71,29 +72,32 @@ class TestWatermodels(object):
-            yield has_identifier, identifier
-            yield itp_in_top, identifier
-            yield coordinates_in_top, identifier
+    def test_watermodelsdat(self):
```
Adding parameterized tests to replace yield tests

ALescoulie commented on mdpow/tests/testing_resources/manifest.yml:
```diff
@@ -1,4 +1,4 @@
-path: mdpow/tests/testing_resources/states
+path: /nfs/homes3/alescoul/MDPOW_dev/mdpow/tests/testing_resources/states
```
Was having issues finding path, will need to replace.

ALescoulie commented on mdpow/fep.py:
```diff
@@ -238,22 +253,33 @@ def load(cfg, section):
-        for k,v in self.iteritems():
-            x[k] = copy.deepcopy(v, memo)
-        return x
+        if sys.version_info.major == 3:
```
iteritems() method not available in python 3, Iterable not variable in python 2, would like more elegant solution, but this should function.

ALescoulie commented on mdpow/restart.py:
```diff
@@ -179,6 +186,17 @@ def __init__(self, *args, **kwargs):
+    @staticmethod
+    def get_pickle():
+        if sys.version_info.major == 3:
+            import _pickle as pickle
```
Ugly compatibility code probably could use six

orbeckst replied:
use six if possible — `from six.moves import pickle` probably, see docs and/or try it out interactively

ALescoulie commented on mdpow/config.py:
```diff
@@ -92,35 +92,42 @@
+
 
-def merge_dicts(user, default):
+# TODO test merge_dicts
```
replaced with recursive method for merging dict containing multiple other dict, passed runinput

ALescoulie reviewed (commented):
Current draft for update to python 3. Modified several outdated sections expecially around dictionary syntax. Replaced merge_dicts with recursive function with method not dependent on isiter.  Many changes just adding spacing to follow PEP8. Tests still need to be modified to run generally, at the moment some paths have been set for locations on workstation for sake of simplicity. Some Errors still occur in the analysis tests some of which appear to be associated with numkit integrate. 

Defiantly still a work in progress, hopefully nearing a functioning version.

orbeckst commented on mdpow/tests/test_Gsolv.py:
```diff
@@ -2,26 +2,25 @@
 import numpy as np
+import gromacs
 from gromacs.utilities import in_dir
+from mdpow.fep import *
+from mdpow.equil import *
```
avoid blanket imports — only import what you need or better, use the qualified names as before

orbeckst commented on mdpow/tests/test_analysis.py:
```diff
@@ -4,17 +4,18 @@
 
 from six.moves import cPickle as pickle
 
-import mdpow.fep
```
leave as before with full paths as it makes it MUCH clearer where code comes from

ALescoulie replied:
I'll make that replacement in fep and equil in a new commit.

orbeckst commented on mdpow/tests/test_analysis.py:
```diff
@@ -4,17 +4,18 @@
+import gromacs
 from numpy.testing import assert_array_almost_equal
 
 from six.moves import cPickle as pickle
```
example for pickle use

orbeckst commented on mdpow/tests/test_emin.py:
```diff
@@ -1,10 +1,13 @@
 class TestEnergyMinimization(object):
     
     def setup(self):
+        gromacs.config.set_gmxrc_environment('~/.conda/envs/MDPOW3/bin/GMXRC')
```
should not be needed

ALescoulie replied:
I'll try without it, I was having some issues with it not finding the conda GMXRC, mostly a temporary solution to get tests running so I could evaluate my code. I defiantly need to remove some of the modifications I made while trying to troubleshoot import errors.

orbeckst commented on mdpow/tests/test_fep.py:
```diff
@@ -1,27 +1,27 @@
-import mdpow
-import mdpow.fep
+import gromacs
+from mdpow.config import *
+from mdpow.fep import *
```
keep full imports

orbeckst commented on mdpow/tests/test_fep.py:
```diff
@@ -50,7 +50,8 @@ class TestFEPschedule(object):
     def setup(self):
         # load default bundled configuration
-        self.cfg = mdpow.config.get_configuration()
+        gromacs.config.set_gmxrc_environment('~/.conda/envs/MDPOW3/bin/GMXRC')
```
should not be needed

orbeckst commented on mdpow/tests/test_fep_script.py:
```diff
@@ -3,18 +3,19 @@
+
 class TestFEPScript(object):
     def setup(self):
+        gromacs.config.set_gmxrc_environment('~/.conda/envs/MDPOW3/bin/GMXRC')
```
should not be needed

ALescoulie commented:
@orbeckst I looked at your comments and to address the concerns will add six.moves to the files currently using cPickle or _pickle. I'll also go back and try to reverse some of the changes I made while trying to troubleshoot getting tests to run for me. They were a bit finicky to get working in particular surrounding sourcing GMXRC and importing the modules.

ALescoulie commented:
Just pushed 3 commits addressing @orbeckst comments on the code. Fixed the test imports and switched to six.moves, also removed the set_gmxrc_environment worked without it.

orbeckst commented:
As discussed offline: start from a recent develop (at least 355dbe8ecb5ccd2d4a3390a65c212f470a899df9) and only add code that is directly necessary for Py 2/3 compatibility. I took your changes for the tests (yield replacement and added them to PR #164 e.g. in 86f563829c440929a51e90f051e30cd06ef88509 ).

orbeckst commented:
Close this PR because it is superseded by PR #166.
